lib/MonadicRewrite.thy

```diff
@@ -613,77 +674,227 @@ lemma monadic_rewrite_weaken_failure:
+lemma monadic_rewrite_corres_l_generic:
+  "\<lbrakk> monadic_rewrite F E Q a a';
+     corres_underlying R nf nf' r P P' a' c;
+     F \<longrightarrow> nf \<rbrakk>
+   \<Longrightarrow> corres_underlying R nf nf' r (P and Q) P' a c"
+  by (fastforce simp add: corres_underlying_def monadic_rewrite_def)
 
-lemma monadic_rewrite_corres2:
+lemma monadic_rewrite_corres_l:
   "\<lbrakk> monadic_rewrite False E Q a a';
-     corres_underlying R False F r P P' a' c \<rbrakk>
-   \<Longrightarrow> corres_underlying R False F r (P and Q) P' a c"
-  by (fastforce simp add: corres_underlying_def monadic_rewrite_def)
+     corres_underlying R nf nf' r P P' a' c \<rbrakk>
+   \<Longrightarrow> corres_underlying R nf nf' r (P and Q) P' a c"
+  by (rule monadic_rewrite_corres_l_generic, simp+)
+
+(* The reasoning behind the failure flag logic is as follows:
+   * if we want to assume non-failure in monadic_rewrite (F), then we have to prove non-failure
+     in corres_underlying (nf'')
+   * if the goal requires us to prove non-failure of RHS, then the corres assumption also needs
+     to prove non-failure *)
+lemma monadic_rewrite_corres_r_generic:
+  "\<lbrakk> monadic_rewrite F E Q c' c;
+     corres_underlying R nf nf'' r P P' a c';
+     F \<longrightarrow> nf''; nf' \<longrightarrow> nf'' \<rbrakk>
+   \<Longrightarrow> corres_underlying R nf nf' r P (P' and Q) a c"
+  by (fastforce simp: corres_underlying_def monadic_rewrite_def)
+
+lemma monadic_rewrite_corres_r:
+  "\<lbrakk> monadic_rewrite False True Q c c';
+     corres_underlying R nf nf' r P P' a c' \<rbrakk>
+   \<Longrightarrow> corres_underlying R nf nf' r P (P' and Q) a c"
+  by (erule (1) monadic_rewrite_corres_r_generic[OF monadic_rewrite_sym], simp+)
 
 lemmas corres_gets_the_bind
-  = monadic_rewrite_corres2[OF monadic_rewrite_bind_head[OF monadic_rewrite_gets_the_gets]]
+  = monadic_rewrite_corres_l[OF monadic_rewrite_bind_head[OF monadic_rewrite_gets_the_gets]]
 
 text \<open>Tool integration\<close>
 
 lemma wpc_helper_monadic_rewrite:
   "monadic_rewrite F E Q' m m'
    \<Longrightarrow> wpc_helper (P, P') (Q, {s. Q' s}) (monadic_rewrite F E (\<lambda>s. s \<in> P') m m')"
-  by (auto simp: wpc_helper_def elim!: monadic_rewrite_imp)
+  by (auto simp: wpc_helper_def elim!: monadic_rewrite_guard_imp)
 
 wpc_setup "\<lambda>m. monadic_rewrite F E Q' m m'" wpc_helper_monadic_rewrite
 wpc_setup "\<lambda>m. monadic_rewrite F E Q' (m >>= c) m'" wpc_helper_monadic_rewrite
 
+text \<open>Tactics\<close>
+
+method monadic_rewrite_step =
+  determ \<open>rule monadic_rewrite_bind_tail monadic_rewrite_bindE_tail\<close>
+
+method monadic_rewrite_solve_head methods m =
+  (rule monadic_rewrite_bind_head monadic_rewrite_bindE_head)?,
+  solves \<open>m, (rule monadic_rewrite_refl)?\<close>
+
+(*
+The most common way of performing monadic rewrite is by doing a pass over the LHS/RHS via some
+a setup step, such as monadic_rewrite_gen_asm + monadic_rewrite_trans (or transverse).
```

Comment:
I think this got mixed up a bit during a rewrite, it was originally `some kind of setup step`
```suggestion
The most common way of performing monadic rewrite is by doing a pass over the LHS/RHS via
a setup step, such as monadic_rewrite_gen_asm + monadic_rewrite_trans (or transverse).
```